makefiles/tools/gdb.inc.mk

```diff
@@ -1,7 +1,11 @@
-# new versions of gdb will support all architectures in one binary
-ifeq ($(shell gdb-multiarch -v > /dev/null 2>&1; echo $$?),0)
-export GDB        ?= gdb-multiarch
+export GDBPREFIX    ?= $(PREFIX)
+
+# If the user installed a magic single target GDB rather than just using
+# gdb-multiarch, there typically is a reason for it. E.g. the shitty upstreaming
```

Comment:
I'm not sure whether we should use dirty words in comments. I don't know if it's really a bad upstream policy from Espressif or if it's just too difficult to get changes into the upstream repository. I'm not familiar with the process for GDB.

Reply:
I had excellent experience with a bug I recently stumbled over with GDBs integrated disassembler. The GDB folks was quick to reply, professional and welcoming.

Note that ESPs are the *only* CPUs I am aware of that do not work with `gdb-multiarch`. It works for any ARM version, any RISC-V version, any PowerPC, any MIPS, any z80 / 8080, any MSP430. It works out of the box for strange beasts like s390 / s390x.

I wouldn't be surprised if the ESPs are the only human made MCUsin the universe sold in quantities > 10K currently unsupported by upstream GDB.

The amount of time to invested and pain and frustration tolerance needed to get tools working with ESP is absolutely unparalleled by any other platform.

I do agree that the Xtensa architecture is surprisingly customizable and the effort to implement proper multilib support is likely higher, especially due to the lack of proper classification of variants (e.g. much unlike ARM / RISC-V). Still, not even non-multilib builds of standard tools are compatible with ESPs.

And with the ESP32C3 it is the exact same pain. There is not reason why a magic `riscv32-esp-elf-gcc` is needed, when there are standard RISC-V toolchains in any major distro provided. I also have very little sympathy for Espressif's choice to use a custom interrupt controller interface over the standard CLIC / PLIC every other vendors uses (I think that in addition to the spec being license free there are even license free implementations they could have used).

The issue of this lack of upstream support becomes apparent when looking at the state of the official toolchains: They still ship ancient versions of GCC and newlib. And for ESP8266 it is unlikely that the official toolchain will ever get an update again.

I am pretty sure that Espressif are the ones to blame. But I'll update the documentation anyway.

Reply:
> I do agree that the Xtensa architecture is surprisingly customizable and the effort to implement proper multilib support is likely higher,

Maybe the situation will become better in the future since Espressif seems to be shifting to RISC-V based SoCs according to the news on [hackster.io](https://www.hackster.io/news/espressif-s-teo-swee-ann-confirms-a-shift-to-risc-v-by-default-unless-we-have-some-special-needs-5e365d49bcc5) six months ago. ESP32-C2 and ESP32-C3 were the first RISC-V based SoCs, ESP32-C6 and [ESP32-H2](https://www.espressif.com/en/news/ESP32-H2_Thread_and_Zigbee_certified) will be the next. Unfortunately, they also provide their own toolchains for RISC-V based SoCs.

> And with the ESP32C3 it is the exact same pain. There is not reason why a magic riscv32-esp-elf-gcc is needed, when there are standard RISC-V toolchains in any major distro provided.

I fully agree. When I was experimenting with ESP32-C3 the first time, I tried to use the `riscv-none-embed` toolchain, but I had no success. On one hand because the source code is incompatible with newer GCC versions and we are not responsible to fix it via a bunch of patches. On the other hand because the binary blobs are compiled with Espressif's  toolchain, which is configured differently regarding the retargetable locking and thread model.

> I also have very little sympathy for Espressif's choice to use a custom interrupt controller interface over the standard CLIC / PLIC every other vendors uses.

I also tried to use the existing architecture specific implementation in `cpu/riscv_common`. Espressif's interrupt controller was only one problem. Another problem was that ESP-IDF uses another context frame structure which is not compatible with the context frame in `cpu/riscv_common`.